### ISSUE TITLE
How to best deal with conversions/seralization to other formats that byte[]?

### ISSUE THREAD

**Issue by johanhaleby** [cloudevents/sdk-java#196]
I'm experimenting with using cloudevents as the storage format for an [event sourcing](https://martinfowler.com/eaaDev/EventSourcing.html) library I'm building (you can find the project [here](https://github.com/johanhaleby/occurrent)). However when I upgraded to `io.cloudevents:cloudevents-json-jackson:2.0.0-milestone1` from `1.3.0` I noticed that the "data" field is no longer generic but is now always `byte[]`. I.e. where you previously could do:

```java
builder.withData(Map.of("hello", "World")). ..
```

you now need to convert the map into a `byte[]`:

```java
builder.withData(objectMapper.writeValueAsBytes(Map.of("hello", "World"))). ..
```

Also `EventFormat` has changed to only deal with `byte[]`. 

I actually think I like this change but it comes with drawbacks in some cases. In my case I'm trying to serialize the CloudEvents to be stored in MongoDB. Since MongoDB understands json natively I'd like to convert the `CloudEvent` instance to a model that MongoDB Java Driver understands, for example a [org.bson.Document](https://mongodb.github.io/mongo-java-driver/4.0/apidocs/bson/org/bson/Document.html). Currently, if I want to use the `io.cloudevents.jackson.JsonFormat`, I need to do double encoding:

```java
CloudEvent ce = ..
byte[] serializedCloudEventAsByteArray = jsonEventFormat.serialize(ce);
// I want to store a Document so I need to convert again
String serializedCloudEventAsString = new String(bytes, UTF_8);
Document cloudEventAsDocument = Document.parse(serializedCloudEventAsString);
mongo.insert(cloudEventAsDocument);
```

This involves several unnecessary steps. Since the [Document.parse](https://mongodb.github.io/mongo-java-driver/4.0/apidocs/bson/org/bson/Document.html#parse(java.lang.String)) method only takes  a `String` I need to first create a `String` from the `byte[]` produced by the `EventFormat` and then pass it to `Document.parse` which will _parse_ the JSON string and convert it into a `Document`.

Compare that with using jackson directly where you could simply do:

```java
CloudEvent ce = ..
Document document = objectMapper.convertValue(ce, Document.class);
mongo.insert(cloudEventAsDocument);
```

which is much more efficient but prohibited when using an instance of `EventFormat`. 

**Questions**

How should one think when dealing with issues as the one above? I think the cloud event spec fit well as storage format but the current implementation prohibits efficient mapping to datastore model (`org.bson.Document` in this case). Should you avoid using `EventFormat` in these cases and write custom serialization/deserialization code?

For reference I've actually copied `io.cloudevents.jackson.JsonFormat` into a custom `DocumentFormat` (which cannot implement `EventFormat` since the types match) to demonstrate the point. As you can see I cannot implement `EventFormat` since it requires a byte array. Other than that it's virtually identical to the existing `JsonFormat` but instead of doing `objectMapper.writeValueAsBytes(..)` I use the `objectMapper.convertValue(..)` method.

```java
package se.haleby.occurrent.eventstore.mongodb.converter;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.module.SimpleModule;
import io.cloudevents.CloudEvent;
import io.cloudevents.core.format.EventDeserializationException;
import io.cloudevents.core.format.EventFormat;
import io.cloudevents.core.format.EventSerializationException;
import io.cloudevents.jackson.CloudEventDeserializer;
import io.cloudevents.jackson.CloudEventSerializer;
import org.bson.Document;

public final class DocumentFormat {

    public static final String CONTENT_TYPE = "application/cloudevents+json";

    private final ObjectMapper mapper;
    private final boolean forceDataBase64Serialization;
    private final boolean forceStringSerialization;

    public DocumentFormat(boolean forceDataBase64Serialization, boolean forceStringSerialization) {
        this.mapper = new ObjectMapper();
        this.mapper.registerModule(getCloudEventJacksonModule(forceDataBase64Serialization, forceStringSerialization));
        this.forceDataBase64Serialization = forceDataBase64Serialization;
        this.forceStringSerialization = forceStringSerialization;
    }

    public DocumentFormat() {
        this(false, false);
    }

    /**
     * @return a copy of this JsonFormat that serialize events with json data with Base64 encoding
     */
    public DocumentFormat withForceJsonDataToBase64() {
        return new DocumentFormat(true, this.forceStringSerialization);
    }

    /**
     * @return a copy of this JsonFormat that serialize events with non-json data as string
     */
    public DocumentFormat withForceNonJsonDataToString() {
        return new DocumentFormat(this.forceDataBase64Serialization, true);
    }

    public Document serialize(CloudEvent event) throws EventSerializationException {
        return mapper.convertValue(event, Document.class);
    }

    public CloudEvent deserialize(Document document) throws EventDeserializationException {
        return mapper.convertValue(document, CloudEvent.class);
    }

    public String serializedContentType() {
        return CONTENT_TYPE;
    }

    /**
     * @return a JacksonModule with CloudEvent serializer/deserializer with default values
     */
    public static SimpleModule getCloudEventJacksonModule() {
        return getCloudEventJacksonModule(false, false);
    }

    /**
     * @return a JacksonModule with CloudEvent serializer/deserializer customizing the data serialization.
     * Look at {@link #withForceJsonDataToBase64()} and {@link #withForceNonJsonDataToString()} for more details.
     */
    public static SimpleModule getCloudEventJacksonModule(boolean forceDataBase64Serialization, boolean forceStringSerialization) {
        final SimpleModule ceModule = new SimpleModule("CloudEvent");
        ceModule.addSerializer(CloudEvent.class, new CloudEventSerializer(forceDataBase64Serialization, forceStringSerialization) {
        });
        ceModule.addDeserializer(CloudEvent.class, new CloudEventDeserializer() {
        });
        return ceModule;
    }

    protected static boolean dataIsJsonContentType(String contentType) {
        // If content type, spec states that we should assume is json
        return contentType == null || contentType.startsWith("application/json") || contentType.startsWith("text/json");
    }
}

**Comment by slinkydeveloper** [comment]
Hi @johanhaleby , i think in this issue there are two separate concerns:

1. The fact that `CloudEvent#getData` (and same on builder side) is a `byte[]` only interface. The change from 1.x to 2.x was made to remove jackson from "core" module, but i'm looking for a strategy to allow support storing pojos in `data` field of the event and, with some pluggable mechanism, solve data marshaller/unmarshallers, avoiding the double encoding (I already have some ideas in mind I plan to show soon with a pr).
2. The problem with `EventFormat`. For that specific case, IMHO it's correct for `EventFormat` to perform CloudEvent <-> byte[] translation, since it's exactly how [event format is specified in the spec](https://github.com/cloudevents/spec/blob/v1.0/spec.md#event-format) and how the users expects to use it in the sdk. I think it makes sense for you to create your own event marshaller/unmarshaller, although I expect you could reuse the jackson `CloudEventSerializer` and `CloudEventDeserializer`.

**Comment by johanhaleby** [comment]
> Hi @johanhaleby , i think in this issue there are two separate concerns:
> 
> 1. The fact that `CloudEvent#getData` (and same on builder side) is a `byte[]` only interface. The change from 1.x to 2.x was made to remove jackson from "core" module, but i'm looking for a strategy to allow support storing pojos in `data` field of the event and, with some pluggable mechanism, solve data marshaller/unmarshallers, avoiding the double encoding (I already have some ideas in mind I plan to show soon with a pr).

Sounds hopeful that you're thinking of a strategy to support pojos. Because even if `EventFormat` always returns `byte[]` (which is fair) it would not be possible to use the `api` if you want to avoid double encoding of the _body_ my case (even if I write a custom `CloudEvent` to `Document` marshaller (1). For example if a user of my API has a simple pojo like this:

```java
public class MyClass {

   private String myString;

   // getters and setters...
}

I'd like for the user to be able to do something like this:

```java
MyClass myClass = ..
CloudEventBuilder.v1()..withDataContentType("application/json").withData(new Document("myString", myClass.getMyString())). .. 
```

or even

```java
MyClass myClass = ..
CloudEventBuilder.v1()..withDataContentType("application/json").withData(myClass). .. 
```

Because, in my case, if the user is forced to do this:

```java
MyClass myClass = ..
byte[] serializedData = objectMapper.writeAsBytes(myClass); // (2)
CloudEventBuilder.v1()..withDataContentType("application/json").withData(serializedData). .. 
```

The `Document` generated by (1) would interpret "data" as `byte[]` and thus will generate a MongoDB bson structure like this:

```json
{
	"specversion":"1.0",
	"id":"86282094-5344-4309-932a-129a7774735e",
	"source":"http://myclass",
	"type":"type",
	"subject":"subject",
	"time":"2020-07-10T14:48:23.272Z",
	"datacontenttype":"application/json",
	"data": "{\"myString\":\"some value\"}"
}
```

I.e. you cannot create MonogDB queries for properties in `data` since it's a `String` and not a bson value (it should be represented as `"data": {"myString": "some value"}`). To solve this my custom marshaller would probably need to check the "datacontenttype" and convert  the serialized content from (2) back into a `Document` which would then be serialized again by the MongoDB driver.

But if there will be support for pojos this could be avoided.

**Comment by slinkydeveloper** [comment]
@johanhaleby mind giving a look here? https://github.com/cloudevents/sdk-java/pull/198

**Comment by slinkydeveloper** [comment]
I think this is solved by #250. I'm working on the `JsonCloudEventData`, that could be a good starting point for your event format too
